### ISSUE TITLE
Auto-detect source and reference for pc_align

### ISSUE THREAD

**Issue by oleg-alexandrov** [NeoGeographyToolkit/StereoPipeline#82]
pc_align expects the reference to be denser than the source point cloud, by the very nature of ICP. The user may not always know this, or may not care. Need a way of estimating automatically which of the two point clouds is denser, and use that as a reference. David suggested looking into lasinfo --compute_density (from lastools, not liblas).

**Comment by oleg-alexandrov** [comment]
Autodetecting source and reference for pc align would result in complication when deciding how to apply the transform to cameras to move from source to ref coordinates and vice versa. [doc](https://stereopipeline.readthedocs.io/en/latest/tools/pc_align.html#applying-a-transform-to-cameras). It will also complicate the convention for how to apply a prior transform to cloud ([doc](https://stereopipeline.readthedocs.io/en/latest/tools/pc_align.html#applying-an-initial-transform)).

A lot of infrastructure in pc_align is built around the fact that the first cloud is the reference. The second is the source. Swapping these internally is a source of confusion and will break subsequent logic.

It is also far from clear how to do it properly when the clouds to have approximately same grid size and if the user would appreciate their decision being overwritten.

In other worlds, pc_align is quite a fragile tool as it is, and alignment can be capricious in the best of times. At least things are properly documented and consistent. Once the user internalizes the rules and knows what to watch for, it works. Auto-guessing would add another layer of uncertainty and cognitive overload.

**Comment by oleg-alexandrov** [comment]
Now, we could presmuably quietly simply call internally ICP the opposite way and swap all when done, while keeping the interface, so this would be a hidden implementation detail, but even this seems to be more pain than what it is worth. The user now gets what they ask for. Normally people have a framework for their particular setup as to how they use this tool and why it makes sense for them. 

Us second-guesing it likely is an additional complication with uncertain results, especially when outlier removal also depends on the order of clouds. This would also need to be documented. So likely not worth it.

**Comment by dshean** [comment]
Makes sense.  Better wrappers around pc_align will help.  A lot of the co-registration uncertainty and sensitivity experiments we're doing will be useful.  Open questions about masking, sampling the input clouds/DEMs.  I'm working on a framework that performs many pc_align runs with same inputs, but different subsets of input datasets, which offers uncertainty estimates for the final transformation.  Not something we necessarily want to bake in, but likely useful for others.
